Design/Packaging/Future_packaging_strategy.md

```diff
@@ -0,0 +1,70 @@
+
+### Conda build
+
+We intend to distribute a conda package using the conda-build system, which will allow advanced users to install Mantid using conda. Users will be able to install this packaged version with, 
+```bash
+conda install -c mantid mantid-workbench
+```
+The user will then be able to interact with the package in a manner familiar to any users of other conda packages, e.g. Spyder, Numpy,
+```bash
+conda install -c mantid mantid-workbench // Install conda package
+python -c "import mantid"  // Import mantid
+MantidWorkbench // Launch workbench
+```
+The runtime requirement specified by the conda build package will be used to produce our offline installers, as described below. 
+
+### Producing Offline (None-Conda) packages
+
+On all operating systems we will need to create the directory from which the package will be made, the command to do this will look something like this:
+```bash
+conda create -p package_dir mantid-workbench --copy -c mantid -c conda-forge -y
+```
+
+In the case of a package that was not uploaded to the Anaconda Cloud like the previous command expects, you can do this from a locally constructed package and achieve the same result using this command:
+```bash
+conda create -p package_dir --use-local mantid-workbench --copy -c conda-forge -y
+```
+
+For Linux operating systems I suggest using FPM (`Effin Package Management - https://github.com/jordansissel/fpm) due to it's simplicity and ease of creating .deb and .rpm installers, from one machine, therefore in the future it should be possible to create all Linux distribution packages from one machine. At present it is not planned to use FPM for MacOSX as it doesn't support .dmg app bundles, and the support it does provide for .pkg files on MacOSX is less than ideal for the same reasons we already migrated away from .pkg files in the first place.
+
+For MacOSX, we need to make .dmg files, we can use a very similar strategy to linux here but with hdiutil (Disk Utility built into MacOS used similarly to how FreeCAD has: https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/osx/create_bundle.sh)
+
+For Windows, we intend to continue using NSIS to produce a .exe installer. We evaluated WIX unfortunately the amount of time it would take to not only switch but to learn how to switch from NSIS is not worth the benefits, this isn't out of the question long term. 
+
+
+## An overall look at the process
+Everything will be achieved from a single script that can be worked into CMake or some other method of calling. The packaging script should return either a locally built Conda-build that is then uploaded or not uploaded, depending on this upload status we then create the environment by either downloading the dependencies and mantid workbench, or just the dependencies using a local mantid workbench constructed with Conda-build. I have produced a Sequence Diagram to better explain the suggested process:
```

Comment:
Analogous to packaging for conda (which I understand much better) is `.spec` for generating rpms. CMake can make rpms without much intervention, but it is much better to hand write a `.spec` file that builds using CMake, and configures the install/packaging steps itself. The `.spec` "language" has facilities to make this fairly straightforward and I believe conda does as well. There will be some judgement applied on where to put some parts of the configuration in this process.

Reply:
The likelihood is, given Windows is a pain when it comes to playing nice with cross platform things, that we will have a POSIX script and a Windows scripts, this is implementation detail however. I am against building a big god like script similar to our present jenkins buildscript. However, I do not think this script will get very large a few hundred lines at most, that's with multiple issues that I currently am not aware of.

These scripts could be called via CMake and perform the packaging role for us, it will be a bit of a step back in terms of our current in built CPack/CMake support but there isn't a better solution that I am aware of. The best we can do here, as far as I know, is mimic how we do it now utilizing a bash and batch script.

Reply:
I think we can work out having the current `package` target do what we need so that `cmake --build . --target package` would work out the right thing to do for us? In that respect I'm not sure it's a step back. The CPack generation of debs/rpms has always been just "good-enough" anyway and there is alot of the install packaging handling that will be cleaned up by conda anyway. 

As Sam linked to, [FreeCad](https://github.com/FreeCAD/FreeCAD-Bundle/tree/master/conda) simply has a set of directories for each of the platforms containing the packaging scripts. If needed I think we can extract out common features into "script modules" used by each?

Are people happy that we use `fpm` to produce the rpms/debs rather than writing spec/control files by hand?